From aed6024423436f9cc7887213956c44a4af910c76 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Sat, 22 Nov 2014 23:56:46 -0500
Subject: [md] dm: return earlier from dm_blk_ioctl if target doesn't implement .ioctl

Message-id: <1416700613-25142-7-git-send-email-snitzer@redhat.com>
Patchwork-id: 100842
O-Subject: [RHEL7.1 PATCH 06/13] dm: return earlier from dm_blk_ioctl if target doesn't implement .ioctl
Bugzilla: 1165246
RH-Acked-by: Joe Thornber <thornber@redhat.com>

BZ: 1165246

Upstream linux-dm.git commit 4d341d8216336174d35cd2575b6b9e4267a88ac8
Author: Mike Snitzer <snitzer@redhat.com>
Date:   Sun Nov 16 14:21:47 2014 -0500

    dm: return earlier from dm_blk_ioctl if target doesn't implement .ioctl

    No point checking if the device is suspended if the current target
    doesn't even implement .ioctl

    Signed-off-by: Mike Snitzer <snitzer@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index 3b4df65..bdf680d 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -525,14 +525,15 @@ retry:
   goto out;
 
  tgt = dm_table_get_target(map, 0);
+ if (!tgt->type->ioctl)
+  goto out;
 
  if (dm_suspended_md(md)) {
   r = -EAGAIN;
   goto out;
  }
 
- if (tgt->type->ioctl)
-  r = tgt->type->ioctl(tgt, cmd, arg);
+ r = tgt->type->ioctl(tgt, cmd, arg);
 
 out:
  dm_put_live_table(md, srcu_idx);
-- 
1.7.1